mat64/list_test.go

```diff
@@ -0,0 +1,706 @@
+				{6, 6, 11, 17},
+				{6, 11, 6, 17},
+				{6, 17, 6, 11},
+				{6, 11, 17, 6},
+				{6, 17, 11, 6},
+				{11, 6, 6, 17},
+				{17, 6, 6, 11},
+				{11, 6, 17, 6},
+				{17, 6, 11, 6},
+				{11, 17, 6, 6},
+				{17, 11, 6, 6},
+			} {
+				// Skip the test if the arguments are compile-time provably incorrect
+				// This is either because the matrix types are wrong at compile time
+				// or because the sizes are impossible
+				ok := legalTypes(aMat, bMat)
+				if !ok {
+					continue
+				}
+				if ok := dimsLegal(aMat, test.ar, test.ac); !ok {
+					continue
+				}
+				if ok := dimsLegal(bMat, test.br, test.bc); !ok {
+					continue
+				}
+				a := makeRandOf(aMat, test.ar, test.ac)
+				{
+					ar, ac := a.Dims()
+					if ar != test.ar || ac != test.ac {
+						panic("makeRandOf returns wrong size" + fmt.Sprintf("%T\n", aMat))
+						continue
+					}
+				}
+				b := makeRandOf(bMat, test.br, test.bc)
+				{
+					br, bc := b.Dims()
+					if br != test.br || bc != test.bc {
+						panic("makeRandOf returns wrong size" + fmt.Sprintf("%T\n", bMat))
+						continue
+					}
+				}
+
+				// Compute the true answer if the sizes are legal.
+				legal := legalSize(test.ar, test.ac, test.br, test.bc)
+				var truth Dense
+				if legal {
+					var aDense, bDense Dense
+					aDense.Clone(a)
+					bDense.Clone(b)
+					denseComparison(&truth, &aDense, &bDense)
+				}
+				// Call the method and check the result with all transpose variations.
+				// Do this by explicitly create a transpose, and then calling T().
+				for _, aTrans := range []bool{true, false} {
+					for _, bTrans := range []bool{true, false} {
+						aCopy := makeCopyOf(a)
+						bCopy := makeCopyOf(b)
+						// Set the proper function call
+						var f func(receiver, a, b Matrix)
+						if aTrans {
+							if bTrans {
+								f = func(receiver, a, b Matrix) {
+									aT := makeTransposeCopyOf(a)
+									r, c := a.Dims()
+									tr, tc := aT.Dims()
+									if r != tc || c != tr {
+										panic("makeTransposeCopyOf returns wrong size" + fmt.Sprintf("%T", a))
+									}
+									bT := makeTransposeCopyOf(b)
+									r, c = b.Dims()
+									tr, tc = bT.Dims()
+									if r != tc || c != tr {
+										panic("makeTransposeCopyOf returns wrong size" + fmt.Sprintf("%T", b))
+									}
+									method(receiver, aT.T(), bT.T())
+								}
+							} else {
+								f = func(receiver, a, b Matrix) {
+									aT := makeTransposeCopyOf(a)
+									method(receiver, aT.T(), b)
+								}
+							}
+						} else {
+							if bTrans {
+								f = func(receiver, a, b Matrix) {
+									bT := makeTransposeCopyOf(b)
+									method(receiver, a, bT.T())
+								}
+							} else {
+								f = func(receiver, a, b Matrix) {
+									method(receiver, a, b)
+								}
+							}
+						}
+						// Test the method for a zero-value of the receiver.
+						errStr := fmt.Sprintf("%T(%T, %T), sizes: %#v, atrans %v, btrans %v", receiver, a, b, test, aTrans, bTrans)
```

Comment:
Add the method name into this: "%T.%s(%T, %T)...". This needs to be threaded in as another parameter to testTwoInput, since it's not possible to determine this by reflection AFAICS.